docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,381 @@
+generated, and those which have manual additions. All the conversion functions
+convert to & from the types in the parent `apis/groups` and `apis/groups/v1`
+directories.
+
+### Package: `apis/group/v1/install/defaulting`
+
+This package contains defaulting functions, and if imported, will register them.
+
+The functions apply to only the `group/v1` types, *not* the parent `group/`
+types.
+
+There are two sorts of defaulting functions, those that can be automatically
+written and those that need human attention.
+
+The former will be autogenerated by looking for `// +default=<value>` comments
+preceding the types or their embedded struct members. Similarly, a
+`// +default=custom` marker will force a stub to be generated. Stubs will be
+generated with a `panic("write me!")` marker to ensure that they are populated.
+
+### Package: `apis/group/v1/install/validation`
+
+This package contains validation functions, and if imported, will register them.
+
+The functions apply to only the `group/v1` types, *not* the parent `group/`
+types.
+
+The allowed imports are constrained; in particular, nothing that would let you
+do RPCs to look up other cluster state is allowed. Validation functions validate
+*only* the object, not cluster-wide or other constraints, which must be validate
+elsewhere. (TODO: copy-pasta this 2x above)
+
+There are two sorts of validation functions, those that can be automatically
+written and those that need human attention.
+
+The former will be autogenerated by looking for directives in the comments
+preceding the types or their embedded struct members. Allowed directives will
+include:
+* `// +validate_min=<numeric value>,validate_max=<numeric value>`
+* `// +validate_regexp=<regexp>`
+* `// +validate_exact=<exact value>`
+
+If no validation directive is in the comments, a stub with a
+`panic("write me!")` marker will be generated. To prevent this, a
+`// +no_validation_required` directive may be added to the comment.
+
+### Package: `apis/group/v1beta1`
+
+This package layout is the same as `apis/group/v1`, but it is an older version
+of the API group, which may lack features.
+
+### Package: `apis/group/v2alpha1`
+
+This package layout is the same as `apis/group/v1`, but it is an newer version
+of the API group, which may have extra features.
+
+## Type support package layout
+
+The above is how to arrange the types for a particular API. But we also must
+consider, what Kubernetes packages do you need to import to use the API types?
+
+```
+pkg
+└── typetools
+    ├── convert
+    ├── deep
+    ├── default
+    └── validate
+```
+
+These packages will live in the kubernetes repo (or another repo that the
+kubernetes team will own and maintain). The interface for all the packages is
+similar: there'll be a `Register(...)` function for the install packages to use,
+and some function to perform the action. So, e.g., `convert.Convert(&in, &out)`,
+`bar := deep.Copy(&foo)`, `default.Apply(&object)`,
```

Comment:
`deep.Copy(&foo, &bar)` is great until you don't actually know the concrete type of `foo`, so maybe we should supply both versions?